File: Project_4/BT7/cpp/top_down.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

// Đọc cây từ input dạng danh sách cha-con
void read_tree(int &n, vector<vector<int>> &tree, int &root) {
    cout << "Nhap so dinh n: ";
    cin >> n;
    tree.assign(n + 1, vector<int>());
    vector<bool> is_child(n + 1, false);
    cout << "Nhap " << n << " dong, moi dong: u k v1 v2 ... vk (dinh u co k con):\n";
    for (int i = 1; i <= n; ++i) {
        int u, k;
        cin >> u >> k;
        for (int j = 0; j < k; ++j) {
            int v; cin >> v;
            tree[u].push_back(v);
            is_child[v] = true;
        }
    }
    // Tìm gốc (đỉnh không là con của ai)
    root = 1;
    for (int i = 1; i <= n; ++i) if (!is_child[i]) root = i;
}

// Duyệt top-down thực sự: in các đỉnh theo từng mức (level-order/BFS)
void top_down(const vector<vector<int>> &tree, int root) {
    queue<pair<int, int>> q; // (đỉnh, độ sâu)
    q.push({root, 0});
    int current_depth = 0;
    vector<int> level_nodes;
    cout << "Duyet top-down (cac dinh theo thu tu khong giam cua do sau, trai sang phai):\n";
    while (!q.empty()) {
        int sz = q.size();
        level_nodes.clear();
        for (int i = 0; i < sz; ++i) {
            auto [u, depth] = q.front(); q.pop();
            level_nodes.push_back(u);
            for (int v : tree[u]) q.push({v, depth + 1});
        }
        cout << "Do sau " << current_depth << ": ";
        for (int u : level_nodes) cout << u << ' ';
        cout << '\n';
        current_depth++;
    }
}

int main() {
    int n, root;
    vector<vector<int>> tree;
    read_tree(n, tree, root);
    top_down(tree, root);
    return 0;
}

File: Project_4/BT7/py/top_down.py
def read_tree():
    n = int(input("Nhap so dinh n: "))
    print(f"Nhap {n} dong, moi dong: u k v1 v2 ... vk (dinh u co k con):")
    tree = [[] for _ in range(n + 1)]
    is_child = [False] * (n + 1)
    for _ in range(n):
        parts = list(map(int, input().split()))
        u, k, *children = parts
        if len(children) < k:
            children += list(map(int, input().split()))
        for v in children:
            tree[u].append(v)
            is_child[v] = True
    root = 1
    for i in range(1, n + 1):
        if not is_child[i]:
            root = i
    return n, tree, root

def top_down(tree, root):
    from collections import deque
    q = deque()
    q.append((root, 0))
    current_depth = 0
    print("Duyet top-down (cac dinh theo thu tu khong giam cua do sau, trai sang phai):")
    while q:
        sz = len(q)
        level_nodes = []
        for _ in range(sz):
            u, depth = q.popleft()
            level_nodes.append(u)
            for v in tree[u]:
                q.append((v, depth + 1))
        print(f"Do sau {current_depth}: ", end='')
        for u in level_nodes:
            print(u, end=' ')
        print()
        current_depth += 1

def main():
    n, tree, root = read_tree()
    top_down(tree, root)

if __name__ == "__main__":
    main()
